### PR TITLE
Add iterative shifter

### PR DESCRIPTION
**Pull request by andreasWallner**
Reproduce issue with
`Test/runMain vexiiriscv.tester.TestBench --load-elf ext/NaxSoftware/baremetal/coremark/build/rv32ima/coremark.elf --with-konata --with-wave --with-spike-log --with-rvls-log --decoders=2 --lanes=2 --allow-bypass-from=0 --with-btb --with-late-alu --with-mul --with-div --with-iterative-shift`

### CODE REVIEW

File: src/main/scala/vexiiriscv/execute/IterativeShiftPlugin.scala [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+          IS_W_RIGHT ? srcp.SRC1(31).asSInt.resize(32).resize(64).asBits | srcp.SRC1.asBits
+        else
+          srcp.SRC1.asBits
+
+      val shiftReg = Reg(cloneOf(srcp.SRC1.asBits))
+      val shifted = cloneOf(srcp.SRC1.asBits)
+      val shiftInput = busy ? shiftReg | signExtended
+      val shiftResult = zeroShift ? signExtended | shifted
+
+      shifted := (LEFT & True).mux(
```
Comment (Dolu1990) [inline src/main/scala/vexiiriscv/execute/IterativeShiftPlugin.scala:83]: I'm asking myself if maybe we could do a "smarter" design with the following hardware shifts : 
- \>\> 1
- \>\> 8
- reverse all bits
- shift by zero case being emulated by doing 2 "reverse all bits"

That way we would have as many mux than now, but much high performance for big shifts, as we could compose shifts using the \>\> 8 and \ >\> 1

In a more generic manner, the number of different >> X possible could be a parameter List[Int]

So, the design would be kinda much closer to the actual BarrelShifterPlugin,  but over multiple cycles.

File: src/main/scala/vexiiriscv/execute/IterativeShiftPlugin.scala [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+      val selected = isValid && SEL
+
+      val amplitudeWidth = if(Riscv.XLEN.get == 64) 6 else 5
+      val shamt = srcp.SRC2.resize(amplitudeWidth bit).asUInt
+      val zeroShift = shamt === 0
+
+      val busy = RegInit(False)
+      val amplitudeReg = Reg(cloneOf(shamt))
+      val amplitude = busy ? amplitudeReg | shamt
+      val done = amplitude(4 downto 1) === 0
```
Comment (Dolu1990) [inline src/main/scala/vexiiriscv/execute/IterativeShiftPlugin.scala:70]: only for 32 bits it works ^^

File: src/main/scala/vexiiriscv/tester/TestBench.scala [hunk 3]
```diff
@@ -456,7 +459,7 @@ object TestBench extends App{
     genConfig.includeSimulation
 
     val simConfig = SpinalSimConfig()
-    simConfig.withFstWave
+    simConfig.withVcdWave
```
Comment (Dolu1990) [inline src/main/scala/vexiiriscv/tester/TestBench.scala:462]: VCD wave are very dangerous as a default, as they can easily blow up a hard drive very first
```suggestion
    simConfig.withFstWave
```

File: src/main/scala/vexiiriscv/execute/IterativeShiftPlugin.scala [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+      val muxInputs = mutable.ArrayBuffer[(Bool, Bits, UInt)]()
+      rightShifts.sorted.foreach(n => {
+        val doIt = if(n > 1) amplitude >= n else True
+        val input = (((ARITHMETIC && srcp.SRC1.msb) #* n) ## shiftReg) >> n
+        val ampl = amplitude - n
+        muxInputs.append((doIt, input, ampl))
+      })
+
+      leftShifts.sorted.foreach(n => {
+        // mask away bits shifted into upper 32 bit in case of SLLW instruction
```
Comment (Dolu1990) [inline src/main/scala/vexiiriscv/execute/IterativeShiftPlugin.scala:127]: Is it necessary ? As anyway, the `ifp.signExtend(wb, layer(op), 32)` will erase all upper bits using the bit 31 as bit sign.

File: src/main/scala/vexiiriscv/execute/IterativeShiftPlugin.scala [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+      SHIFT_RESULT := (if (lateResult) shiftReg else muxed)
+    }
+
+    val format = new el.Execute(formatAt) {
+      wb.valid := SEL
+      wb.payload := SHIFT_RESULT
+      if(Riscv.XLEN.get == 64) {
+        when(ARITHMETIC & IS_W) {
+          wb.payload(32, 32 bit).setAllTo(SHIFT_RESULT(31))
```
Comment (Dolu1990) [inline src/main/scala/vexiiriscv/execute/IterativeShiftPlugin.scala:196]: This seems redondant with the `ifp.signExtend(wb, layer(op), 32)` 

ifp => IntFormatPlugin => a plugin which provide a integer writeback path with optional static sign/zero extends

Comment (andreasWallner) [reply src/main/scala/vexiiriscv/execute/IterativeShiftPlugin.scala:196]: right, done^

File: src/main/scala/vexiiriscv/tester/TestBench.scala (hunk 3 above)
Comment (Dolu1990) [reply src/main/scala/vexiiriscv/tester/TestBench.scala]: I guess overall, we would need to add all of those through argument / conifiguration / envrionnement variables ^^

I also have a lot the habit to modify / comment code in order to fit my specific needs, but overall, that doesn't scale up well XD

Comment (andreasWallner) [reply src/main/scala/vexiiriscv/tester/TestBench.scala]: that's why I often use this one: https://github.com/andreasWallner/spinalStuff/blob/90604fce38fab63d7a366e91a158b6e6f03e56e7/src/main/scala/andreasWallner/sim/package.scala#L268 ;-)
(though I'm not that happy with the name anymore)

File: src/main/scala/vexiiriscv/execute/IterativeShiftPlugin.scala (hunk 4 above)
Comment (andreasWallner) [reply src/main/scala/vexiiriscv/execute/IterativeShiftPlugin.scala]: fixed

**Comment by Dolu1990** [comment]
Thanks :)

**Pull request merged**